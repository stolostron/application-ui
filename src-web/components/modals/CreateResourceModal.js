/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import R from 'ramda'
import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import jsYaml from 'js-yaml'
import {
  Button,
  Icon,
  InlineNotification,
  Loading,
  ComposedModal,
  ModalHeader,
  ModalBody,
  ModalFooter,
  Tabs,
  Tab
} from 'carbon-components-react'
import resources from '../../../lib/shared/resources'
import msgs from '../../../nls/platform.properties'
import getResourceDefinitions from '../../definitions'
import YamlEditor from '../common/YamlEditor'

resources(() => {
  require('../../../scss/modal.scss')
})

const initialState = {
  modalOpen: false,
  processing: false,
  yaml: '',
  yamlParsingError: null,
  createError: null,
  dirty: false,
  sample: null
}

// A hacky way to delay making the fetch call
const waitTime = ms => {
  const start = new Date().getTime()
  let end = start
  while (end < start + ms) {
    end = new Date().getTime()
  }
}

class CreateResourceModal extends React.PureComponent {
  static propTypes = {
    headingTextKey: PropTypes.string,
    helpLink: PropTypes.string,
    iconDescription: PropTypes.string,
    onCreateResource: PropTypes.func,
    onSubmitFunction: PropTypes.func,
    resourceDescriptionKey: PropTypes.string,
    resourceType: PropTypes.object,
    resourceTypeName: PropTypes.string,
    sampleContent: PropTypes.array,
    sampleTabs: PropTypes.object
  };

  state = initialState;

  handleModalOpen = () => {
    this.setState({ modalOpen: true })
  };
  handleModalCancel = () => {
    this.setState(initialState)
  };
  handleModalSubmit = () => {
    let resources
    try {
      // the next line code will split the yaml content into multi-parts
      // if '---' found in the content
      resources = jsYaml.safeLoadAll(this.state.yaml)
    } catch (e) {
      this.setState({ yamlParsingError: e })
      return
    }
    this.setState({ yamlParsingError: null, processing: true })

    this.props.onCreateResource(resources).then(result => {
      // errors can be in createApplication and createResources depending on type of resource
      const errors = R.concat(
        R.pathOr([], ['data', 'createApplication', 'errors'], result),
        R.pathOr([], ['data', 'createResources', 'errors'], result)
      )

      if (errors && errors.length > 0) {
        this.setState({
          createError: {
            message: errors[0].message
          },
          processing: false
        })
      } else {
        this.setState(initialState)
        // If there is a on Submit function passed in we want to execute it
        if (this.props.onSubmitFunction) {
          waitTime(7000)
          this.props.onSubmitFunction()
        }
      }
    })
  };

  handleEditorChange = yaml => {
    this.setState({ yaml, dirty: true })
  };

  handleParsingError = yamlParsingError => this.setState({ yamlParsingError });

  handleNotificationClosed = () => this.setState({ yamlParsingError: null });

  isSubmitDisabled = () => this.state.processing === true;

  render() {
    const { resourceType } = this.props
    const { validator } = getResourceDefinitions(resourceType)
    const tabs = this.props.sampleTabs
    const tabsSampleContent = this.props.sampleContent
    const tabsHandleEditorChange = this.handleEditorChange
    const tabsHandleParsingError = this.handleParsingError
    const tabsYaml = this.state.yaml
    const errorMsg = this.state.createError && this.state.createError.message
    return (
      <div>
        <Button
          icon="add--glyph"
          small
          id={msgs.get(this.props.resourceTypeName, this.context.locale)}
          iconDescription={this.props.iconDescription}
          key="create-resource"
          onClick={this.handleModalOpen}
        >
          {msgs.get(this.props.resourceTypeName, this.context.locale)}
        </Button>
        {this.state.modalOpen && (
          <ComposedModal
            className="modal-with-editor"
            open={this.state.modalOpen}
            role="region"
            onClose={() => false}
          >
            <ModalHeader
              title={msgs.get(this.props.headingTextKey, this.context.locale)}
              buttonOnClick={this.handleModalCancel}
            />
            <ModalBody>
              <div className="bx--modal-content-desc">
                <div className="yaml-instructions">
                  {msgs.get(
                    this.props.resourceDescriptionKey,
                    this.context.locale
                  )}
                </div>

                {this.props.helpLink && (
                  <div className="help-link">
                    <a href={this.props.helpLink} target="_blank">
                      {msgs.get('link.help.writing', this.context.locale)}
                    </a>

                    <a href={this.props.helpLink} target="_blank">
                      <Icon
                        name="icon--launch"
                        fill="#6089bf"
                        description=""
                        className="helpLinkIcon"
                      />
                    </a>
                  </div>
                )}
              </div>
              {this.state.dirty &&
                this.state.yamlParsingError && (
                  <InlineNotification
                    kind="error"
                    title={msgs.get('error.parse', this.context.locale)}
                    iconDescription=""
                    subtitle={msgs.get(
                      'error.parse.reason',
                      this.context.locale
                    )}
                    onCloseButtonClick={this.handleNotificationClosed}
                  />
              )}
              {(this.state.createError || errorMsg) && (
                <InlineNotification
                  kind="error"
                  title={msgs.get('error.create', this.context.locale)}
                  iconDescription=""
                  // show default msg if errorMsg is not set
                  subtitle={
                    errorMsg ||
                    msgs.get('error.create.reason', this.context.locale)
                  }
                  onCloseButtonClick={this.handleNotificationClosed}
                />
              )}
              {this.props.sampleTabs ? (
                <div className="yamlSampleTabsContainer">
                  <Tabs
                    className={
                      !this.state.dirty
                        ? 'yamlSampleTabs'
                        : 'yamlSampleTabs hidden'
                    }
                  >
                    {Object.keys(tabs).map((key, i) => {
                      return (
                        <Tab
                          onClick={() => {
                            this.setState({ sample: tabsSampleContent[i] })
                          }}
                          onKeyDown={() => {
                            // noop function for optional property
                          }}
                          key={tabs[key]}
                          label={tabs[key]}
                        />
                      )
                    })}
                  </Tabs>
                  <YamlEditor
                    validator={validator}
                    onYamlChange={tabsHandleEditorChange}
                    handleParsingError={tabsHandleParsingError}
                    yaml={
                      this.state.dirty
                        ? tabsYaml
                        : this.state.sample
                          ? this.state.sample
                          : tabsSampleContent[0]
                    }
                  />
                </div>
              ) : (
                <YamlEditor
                  validator={validator}
                  onYamlChange={tabsHandleEditorChange}
                  handleParsingError={this.handleParsingError}
                  yaml={
                    this.state.dirty ? this.state.yaml : tabsSampleContent[0]
                  }
                />
              )}
              {this.state.processing && <Loading />}
            </ModalBody>
            <ModalFooter>
              <div id="modalFotterBtnDiv">
                <Button
                  className="bx--btn--secondary"
                  type="button"
                  onClick={this.handleModalCancel}
                >
                  {msgs.get('actions.cancel', this.context.locale)}
                </Button>
                <Button
                  type="button"
                  disabled={this.isSubmitDisabled()}
                  onClick={this.handleModalSubmit}
                >
                  {msgs.get('modal.button.save', this.context.locale)}
                </Button>
              </div>
            </ModalFooter>
          </ComposedModal>
        )}
      </div>
    )
  }
}

CreateResourceModal.contextType = {
  locale: PropTypes.locale
}

const mapDispatchToProps = (dispatch, { onCreateResource }) => {
  return {
    onCreateResource: yaml => onCreateResource(dispatch, yaml)
  }
}

export default connect(() => ({}), mapDispatchToProps)(CreateResourceModal)
