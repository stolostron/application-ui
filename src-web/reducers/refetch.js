/*******************************************************************************
 * Copyright (c) 2020 Red Hat, Inc.
 *******************************************************************************/

import { REFETCH_INTERVAL_UPDATE } from '../actions'

export const refetch = (state, action) => {
  switch (action.type) {
  case REFETCH_INTERVAL_UPDATE:
    return action.data
  default:
    // get the poll interval from RefreshTimeSelect component
    return Object.assign({}, state, {})
  }
}
