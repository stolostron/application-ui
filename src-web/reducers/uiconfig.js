/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

import { UICONFIG_RECEIVE_SUCCESS } from '../actions'

export const uiconfig = (state = null, action) => {
  switch (action.type) {
  case UICONFIG_RECEIVE_SUCCESS:
    return action.data
  default:
    return state
  }
}
