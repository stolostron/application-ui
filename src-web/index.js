/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
import React from 'react'
import { hydrate } from 'react-dom'
import { Provider } from 'react-redux'
import thunkMiddleware from 'redux-thunk'
import { createLogger } from 'redux-logger'
import { createStore, combineReducers, applyMiddleware, compose } from 'redux'
import { Router } from 'react-router-dom'
import { ApolloProvider } from 'react-apollo'
import App from './containers/App'
import ScrollToTop from './components/common/ScrollToTop'
import * as reducers from './reducers'
import config from '../lib/shared/config'
import apolloClient from '../lib/client/apollo-client'
import history from 'history'

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose

window.SHARED_HISTORY = window.SHARED_HISTORY
  ? window.SHARED_HISTORY
  : history.createBrowserHistory()

const loggerMiddleware = createLogger()
// Grab the state from a global variable injected into the server-generated HTML
const preloadedState = window.__PRELOADED_STATE__

const middleware = [thunkMiddleware] // lets us dispatch() functions
if (
  !window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
  config['featureFlags:reduxLogger']
) {
  middleware.push(loggerMiddleware) // middleware that logs actions
}

// Allow the passed state to be garbage-collected
delete window.__PRELOADED_STATE__

// Create Redux store with initial state
const store = createStore(
  combineReducers(reducers),
  preloadedState,
  composeEnhancers(applyMiddleware(...middleware))
)

hydrate(
  <ApolloProvider client={apolloClient.getClient()}>
    <Provider store={store}>
      <Router history={window.SHARED_HISTORY}>
        <ScrollToTop>
          <App />
        </ScrollToTop>
      </Router>
    </Provider>
  </ApolloProvider>,
  document.getElementById('page')
)
