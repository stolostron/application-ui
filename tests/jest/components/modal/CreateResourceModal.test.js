/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
"use strict";
jest.mock("../../../../lib/client/apollo-client", () => ({
  getClient: jest.fn(() => {
    return null;
  }),
  updateResource: jest.fn(
    (resourceType, namespace, name, body, selfLink, resourcePath) => {
      return Promise.resolve({});
    }
  ),
  getResource: jest.fn((resourceType, variables) => {
    const data = {
      data: {
        items: [
          {
            metadata: {
              creationTimestamp: "2020-04-06T22:27:05Z",
              generation: 2,
              name: "guestbook-app",
              namespace: "default",
              resourceVersion: "840144",
              selfLink:
                "/apis/app.k8s.io/v1beta1/namespaces/default/applications/guestbook-app",
              uid: "0221dae9-b6b9-40cb-8cba-473011a750e0"
            },
            raw: {
              apiVersion: "app.k8s.io/v1beta1",
              kind: "Application",
              metadata: {
                creationTimestamp: "2020-04-06T22:27:05Z",
                generation: 2,
                name: "guestbook-app",
                namespace: "default",
                resourceVersion: "840144",
                selfLink:
                  "/apis/app.k8s.io/v1beta1/namespaces/default/applications/guestbook-app",
                uid: "0221dae9-b6b9-40cb-8cba-473011a750e0"
              }
            }
          }
        ]
      }
    };

    return Promise.resolve(data);
  })
}));
import React from "react";

import CreateResourceModal from "../../../../src-web/components/modals/CreateResourceModal";
import { mount } from "enzyme";
import * as reducers from "../../../../src-web/reducers";
import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunkMiddleware from "redux-thunk";
import { resourceModalLabels } from "./ModalsTestingData";
import toJson from "enzyme-to-json";
import { BrowserRouter } from "react-router-dom";

describe("CreateResourceModal test", () => {
  const handleModalClose = jest.fn();
  const handleModalSubmit = jest.fn();
  const resourceType = { name: "HCMApplication", list: "HCMApplicationList" };
  const resourceTypeName = "description.application";
  const preloadedState = window.__PRELOADED_STATE__;
  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  const middleware = [thunkMiddleware];
  const store = createStore(
    combineReducers(reducers),
    preloadedState,
    composeEnhancers(applyMiddleware(...middleware))
  );

  const data = {
    clusterName: "",
    kind: "",
    name: "guestbook-app",
    namespace: "default",
    selfLink:
      "/apis/app.k8s.io/v1beta1/namespaces/default/applications/guestbook-app",
    _uid: "0221dae9-b6b9-40cb-8cba-473011a750e0"
  };

  it("renders as expected with error", () => {
    const component = mount(
      <BrowserRouter>
        <CreateResourceModal
          handleClose={handleModalClose}
          handleSubmit={handleModalSubmit}
          label={resourceModalLabels}
          locale={"en"}
          open={true}
          resourceType={resourceType}
          resourceTypeName={resourceTypeName}
          store={store}
          data={data}
          sampleContent={["aaaaa"]}
        />
      </BrowserRouter>
    );
    expect(toJson(component.instance())).toMatchSnapshot();
    expect(toJson(component.update())).toMatchSnapshot();
    expect(toJson(component)).toMatchSnapshot();

    component
      .find(".bx--btn--primary")
      .at(0)
      .simulate("click");
  });

  store.dirty = true;

  it("renders as expected 1", () => {
    const component = mount(
      <BrowserRouter>
        <CreateResourceModal
          handleClose={handleModalClose}
          handleSubmit={handleModalSubmit}
          label={resourceModalLabels}
          locale={"en"}
          open={true}
          resourceType={resourceType}
          resourceTypeName={resourceTypeName}
          store={store}
          data={data}
          sampleContent={["aaaaa"]}
        />
      </BrowserRouter>
    );
    expect(toJson(component.instance())).toMatchSnapshot();
    expect(toJson(component.update())).toMatchSnapshot();
    expect(toJson(component)).toMatchSnapshot();

    component
      .find(".bx--btn--primary")
      .at(0)
      .simulate("click");
  });
});
