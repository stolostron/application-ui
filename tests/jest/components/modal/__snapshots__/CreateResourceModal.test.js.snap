// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateResourceModal test renders as expected 1 1`] = `undefined`;

exports[`CreateResourceModal test renders as expected 1 2`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(CreateResourceModal)
      data={
        Object {
          "_uid": "0221dae9-b6b9-40cb-8cba-473011a750e0",
          "clusterName": "",
          "kind": "",
          "name": "guestbook-app",
          "namespace": "default",
          "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/default/applications/guestbook-app",
        }
      }
      handleClose={[MockFunction]}
      handleSubmit={[MockFunction]}
      label={
        Object {
          "heading": "modal.remove-hcmapplication.heading",
          "label": "modal.remove-hcmapplication.label",
          "primaryBtn": "modal.remove-hcmapplication.heading",
        }
      }
      locale="en"
      open={true}
      resourceType={
        Object {
          "list": "HCMApplicationList",
          "name": "HCMApplication",
        }
      }
      sampleContent={
        Array [
          "aaaaa",
        ]
      }
      store={
        Object {
          "dirty": true,
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <CreateResourceModal
        data={
          Object {
            "_uid": "0221dae9-b6b9-40cb-8cba-473011a750e0",
            "clusterName": "",
            "kind": "",
            "name": "guestbook-app",
            "namespace": "default",
            "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/default/applications/guestbook-app",
          }
        }
        deleteSuccessFinished={[Function]}
        handleClose={[MockFunction]}
        handleSubmit={[MockFunction]}
        label={
          Object {
            "heading": "modal.remove-hcmapplication.heading",
            "label": "modal.remove-hcmapplication.label",
            "primaryBtn": "modal.remove-hcmapplication.heading",
          }
        }
        locale="en"
        mutateSuccessFinished={[Function]}
        onCreateResource={[Function]}
        open={true}
        resourceType={
          Object {
            "list": "HCMApplicationList",
            "name": "HCMApplication",
          }
        }
        sampleContent={
          Array [
            "aaaaa",
          ]
        }
        store={
          Object {
            "dirty": true,
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
        storeSubscription={
          Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "dirty": true,
              "dispatch": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
              Symbol(observable): [Function],
            },
            "unsubscribe": [Function],
          }
        }
      >
        <div>
          <Button
            disabled={false}
            icon="add--glyph"
            iconDescription="Provide icon description if icon is used"
            key="create-resource"
            kind="primary"
            onClick={[Function]}
            small={true}
            tabIndex={0}
            type="button"
          >
            <button
              className="bx--btn bx--btn--sm bx--btn--primary"
              disabled={false}
              onClick={[Function]}
              tabIndex={0}
              type="button"
            >
              <Icon
                className="bx--btn__icon"
                description="Provide icon description if icon is used"
                fillRule="evenodd"
                name="add--glyph"
                role="img"
              >
                <svg
                  alt="Provide icon description if icon is used"
                  aria-label="Provide icon description if icon is used"
                  className="bx--btn__icon"
                  fillRule="evenodd"
                  height="16"
                  name="add--glyph"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <title>
                    Provide icon description if icon is used
                  </title>
                  <path
                    d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm4 9H9v3H7V9H4V7h3V4h2v3h3v2z"
                    key="key0"
                  />
                </svg>
              </Icon>
            </button>
          </Button>
        </div>
      </CreateResourceModal>
    </Connect(CreateResourceModal)>
  </Router>
</BrowserRouter>
`;

exports[`CreateResourceModal test renders as expected 1 3`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(CreateResourceModal)
      data={
        Object {
          "_uid": "0221dae9-b6b9-40cb-8cba-473011a750e0",
          "clusterName": "",
          "kind": "",
          "name": "guestbook-app",
          "namespace": "default",
          "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/default/applications/guestbook-app",
        }
      }
      handleClose={[MockFunction]}
      handleSubmit={[MockFunction]}
      label={
        Object {
          "heading": "modal.remove-hcmapplication.heading",
          "label": "modal.remove-hcmapplication.label",
          "primaryBtn": "modal.remove-hcmapplication.heading",
        }
      }
      locale="en"
      open={true}
      resourceType={
        Object {
          "list": "HCMApplicationList",
          "name": "HCMApplication",
        }
      }
      sampleContent={
        Array [
          "aaaaa",
        ]
      }
      store={
        Object {
          "dirty": true,
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <CreateResourceModal
        data={
          Object {
            "_uid": "0221dae9-b6b9-40cb-8cba-473011a750e0",
            "clusterName": "",
            "kind": "",
            "name": "guestbook-app",
            "namespace": "default",
            "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/default/applications/guestbook-app",
          }
        }
        deleteSuccessFinished={[Function]}
        handleClose={[MockFunction]}
        handleSubmit={[MockFunction]}
        label={
          Object {
            "heading": "modal.remove-hcmapplication.heading",
            "label": "modal.remove-hcmapplication.label",
            "primaryBtn": "modal.remove-hcmapplication.heading",
          }
        }
        locale="en"
        mutateSuccessFinished={[Function]}
        onCreateResource={[Function]}
        open={true}
        resourceType={
          Object {
            "list": "HCMApplicationList",
            "name": "HCMApplication",
          }
        }
        sampleContent={
          Array [
            "aaaaa",
          ]
        }
        store={
          Object {
            "dirty": true,
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
        storeSubscription={
          Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "dirty": true,
              "dispatch": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
              Symbol(observable): [Function],
            },
            "unsubscribe": [Function],
          }
        }
      >
        <div>
          <Button
            disabled={false}
            icon="add--glyph"
            iconDescription="Provide icon description if icon is used"
            key="create-resource"
            kind="primary"
            onClick={[Function]}
            small={true}
            tabIndex={0}
            type="button"
          >
            <button
              className="bx--btn bx--btn--sm bx--btn--primary"
              disabled={false}
              onClick={[Function]}
              tabIndex={0}
              type="button"
            >
              <Icon
                className="bx--btn__icon"
                description="Provide icon description if icon is used"
                fillRule="evenodd"
                name="add--glyph"
                role="img"
              >
                <svg
                  alt="Provide icon description if icon is used"
                  aria-label="Provide icon description if icon is used"
                  className="bx--btn__icon"
                  fillRule="evenodd"
                  height="16"
                  name="add--glyph"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <title>
                    Provide icon description if icon is used
                  </title>
                  <path
                    d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm4 9H9v3H7V9H4V7h3V4h2v3h3v2z"
                    key="key0"
                  />
                </svg>
              </Icon>
            </button>
          </Button>
        </div>
      </CreateResourceModal>
    </Connect(CreateResourceModal)>
  </Router>
</BrowserRouter>
`;

exports[`CreateResourceModal test renders as expected with error 1`] = `undefined`;

exports[`CreateResourceModal test renders as expected with error 2`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(CreateResourceModal)
      data={
        Object {
          "_uid": "0221dae9-b6b9-40cb-8cba-473011a750e0",
          "clusterName": "",
          "kind": "",
          "name": "guestbook-app",
          "namespace": "default",
          "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/default/applications/guestbook-app",
        }
      }
      handleClose={[MockFunction]}
      handleSubmit={[MockFunction]}
      label={
        Object {
          "heading": "modal.remove-hcmapplication.heading",
          "label": "modal.remove-hcmapplication.label",
          "primaryBtn": "modal.remove-hcmapplication.heading",
        }
      }
      locale="en"
      open={true}
      resourceType={
        Object {
          "list": "HCMApplicationList",
          "name": "HCMApplication",
        }
      }
      sampleContent={
        Array [
          "aaaaa",
        ]
      }
      store={
        Object {
          "dirty": true,
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <CreateResourceModal
        data={
          Object {
            "_uid": "0221dae9-b6b9-40cb-8cba-473011a750e0",
            "clusterName": "",
            "kind": "",
            "name": "guestbook-app",
            "namespace": "default",
            "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/default/applications/guestbook-app",
          }
        }
        deleteSuccessFinished={[Function]}
        handleClose={[MockFunction]}
        handleSubmit={[MockFunction]}
        label={
          Object {
            "heading": "modal.remove-hcmapplication.heading",
            "label": "modal.remove-hcmapplication.label",
            "primaryBtn": "modal.remove-hcmapplication.heading",
          }
        }
        locale="en"
        mutateSuccessFinished={[Function]}
        onCreateResource={[Function]}
        open={true}
        resourceType={
          Object {
            "list": "HCMApplicationList",
            "name": "HCMApplication",
          }
        }
        sampleContent={
          Array [
            "aaaaa",
          ]
        }
        store={
          Object {
            "dirty": true,
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
        storeSubscription={
          Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "dirty": true,
              "dispatch": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
              Symbol(observable): [Function],
            },
            "unsubscribe": [Function],
          }
        }
      >
        <div>
          <Button
            disabled={false}
            icon="add--glyph"
            iconDescription="Provide icon description if icon is used"
            key="create-resource"
            kind="primary"
            onClick={[Function]}
            small={true}
            tabIndex={0}
            type="button"
          >
            <button
              className="bx--btn bx--btn--sm bx--btn--primary"
              disabled={false}
              onClick={[Function]}
              tabIndex={0}
              type="button"
            >
              <Icon
                className="bx--btn__icon"
                description="Provide icon description if icon is used"
                fillRule="evenodd"
                name="add--glyph"
                role="img"
              >
                <svg
                  alt="Provide icon description if icon is used"
                  aria-label="Provide icon description if icon is used"
                  className="bx--btn__icon"
                  fillRule="evenodd"
                  height="16"
                  name="add--glyph"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <title>
                    Provide icon description if icon is used
                  </title>
                  <path
                    d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm4 9H9v3H7V9H4V7h3V4h2v3h3v2z"
                    key="key0"
                  />
                </svg>
              </Icon>
            </button>
          </Button>
        </div>
      </CreateResourceModal>
    </Connect(CreateResourceModal)>
  </Router>
</BrowserRouter>
`;

exports[`CreateResourceModal test renders as expected with error 3`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(CreateResourceModal)
      data={
        Object {
          "_uid": "0221dae9-b6b9-40cb-8cba-473011a750e0",
          "clusterName": "",
          "kind": "",
          "name": "guestbook-app",
          "namespace": "default",
          "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/default/applications/guestbook-app",
        }
      }
      handleClose={[MockFunction]}
      handleSubmit={[MockFunction]}
      label={
        Object {
          "heading": "modal.remove-hcmapplication.heading",
          "label": "modal.remove-hcmapplication.label",
          "primaryBtn": "modal.remove-hcmapplication.heading",
        }
      }
      locale="en"
      open={true}
      resourceType={
        Object {
          "list": "HCMApplicationList",
          "name": "HCMApplication",
        }
      }
      sampleContent={
        Array [
          "aaaaa",
        ]
      }
      store={
        Object {
          "dirty": true,
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <CreateResourceModal
        data={
          Object {
            "_uid": "0221dae9-b6b9-40cb-8cba-473011a750e0",
            "clusterName": "",
            "kind": "",
            "name": "guestbook-app",
            "namespace": "default",
            "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/default/applications/guestbook-app",
          }
        }
        deleteSuccessFinished={[Function]}
        handleClose={[MockFunction]}
        handleSubmit={[MockFunction]}
        label={
          Object {
            "heading": "modal.remove-hcmapplication.heading",
            "label": "modal.remove-hcmapplication.label",
            "primaryBtn": "modal.remove-hcmapplication.heading",
          }
        }
        locale="en"
        mutateSuccessFinished={[Function]}
        onCreateResource={[Function]}
        open={true}
        resourceType={
          Object {
            "list": "HCMApplicationList",
            "name": "HCMApplication",
          }
        }
        sampleContent={
          Array [
            "aaaaa",
          ]
        }
        store={
          Object {
            "dirty": true,
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
        }
        storeSubscription={
          Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "dirty": true,
              "dispatch": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
              Symbol(observable): [Function],
            },
            "unsubscribe": [Function],
          }
        }
      >
        <div>
          <Button
            disabled={false}
            icon="add--glyph"
            iconDescription="Provide icon description if icon is used"
            key="create-resource"
            kind="primary"
            onClick={[Function]}
            small={true}
            tabIndex={0}
            type="button"
          >
            <button
              className="bx--btn bx--btn--sm bx--btn--primary"
              disabled={false}
              onClick={[Function]}
              tabIndex={0}
              type="button"
            >
              <Icon
                className="bx--btn__icon"
                description="Provide icon description if icon is used"
                fillRule="evenodd"
                name="add--glyph"
                role="img"
              >
                <svg
                  alt="Provide icon description if icon is used"
                  aria-label="Provide icon description if icon is used"
                  className="bx--btn__icon"
                  fillRule="evenodd"
                  height="16"
                  name="add--glyph"
                  role="img"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <title>
                    Provide icon description if icon is used
                  </title>
                  <path
                    d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm4 9H9v3H7V9H4V7h3V4h2v3h3v2z"
                    key="key0"
                  />
                </svg>
              </Icon>
            </button>
          </Button>
        </div>
      </CreateResourceModal>
    </Connect(CreateResourceModal)>
  </Router>
</BrowserRouter>
`;
